{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5219755a",
   "metadata": {},
   "source": [
    "# Genotipador versão 000.01\n",
    "Inicio do desenvolvimento 2021/09/06\n",
    "Magalhães, RDM."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "376a2e57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "344.0\n",
      "750.0\n",
      "[0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0]\n",
      "   aa_pos aa_WT aa_Mut aa_Fd aa_Ch\n",
      "0     346     R      K     R     0\n",
      "1     417     K      T     K     0\n",
      "2     417     K      N     K     0\n",
      "3     452     L      Q     R     0\n",
      "4     452     L      R     R     1\n",
      "5     477     S      N     S     0\n",
      "6     478     T      K     K     1\n",
      "7     484     E      K     E     0\n",
      "8     484     E      Q     E     0\n",
      "9     490     F      S     F     0\n",
      "10    501     N      Y     N     0\n",
      "11    614     D      G     G     1\n",
      "12    655     H      Y     H     0\n",
      "13    677     Q      H     Q     0\n",
      "14    681     P      R     R     1\n",
      "15    681     P      H     R     0\n",
      "16    701     A      V     A     0\n",
      "FN8863\n",
      "Delta(B.1.617.2)\n",
      "\n",
      "\n",
      "342.0\n",
      "751.0\n",
      "[0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0]\n",
      "   aa_pos aa_WT aa_Mut aa_Fd aa_Ch\n",
      "0     346     R      K     R     0\n",
      "1     417     K      T     K     0\n",
      "2     417     K      N     K     0\n",
      "3     452     L      Q     R     0\n",
      "4     452     L      R     R     1\n",
      "5     477     S      N     S     0\n",
      "6     478     T      K     K     1\n",
      "7     484     E      K     E     0\n",
      "8     484     E      Q     E     0\n",
      "9     490     F      S     F     0\n",
      "10    501     N      Y     N     0\n",
      "11    614     D      G     G     1\n",
      "12    655     H      Y     H     0\n",
      "13    677     Q      H     Q     0\n",
      "14    681     P      R     R     1\n",
      "15    681     P      H     R     0\n",
      "16    701     A      V     A     0\n",
      "FN8920\n",
      "Delta(B.1.617.2)\n",
      "\n",
      "\n",
      "344.0\n",
      "751.0\n",
      "[0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0]\n",
      "   aa_pos aa_WT aa_Mut aa_Fd aa_Ch\n",
      "0     346     R      K     R     0\n",
      "1     417     K      T     K     0\n",
      "2     417     K      N     K     0\n",
      "3     452     L      Q     R     0\n",
      "4     452     L      R     R     1\n",
      "5     477     S      N     S     0\n",
      "6     478     T      K     K     1\n",
      "7     484     E      K     E     0\n",
      "8     484     E      Q     E     0\n",
      "9     490     F      S     F     0\n",
      "10    501     N      Y     N     0\n",
      "11    614     D      G     G     1\n",
      "12    655     H      Y     H     0\n",
      "13    677     Q      H     Q     0\n",
      "14    681     P      R     R     1\n",
      "15    681     P      H     R     0\n",
      "16    701     A      V     A     0\n",
      "FN8941\n",
      "Delta(B.1.617.2)\n",
      "\n",
      "\n",
      "340.0\n",
      "748.0\n",
      "[0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0]\n",
      "   aa_pos aa_WT aa_Mut aa_Fd aa_Ch\n",
      "0     346     R      K     R     0\n",
      "1     417     K      T     K     0\n",
      "2     417     K      N     K     0\n",
      "3     452     L      Q     R     0\n",
      "4     452     L      R     R     1\n",
      "5     477     S      N     S     0\n",
      "6     478     T      K     K     1\n",
      "7     484     E      K     E     0\n",
      "8     484     E      Q     E     0\n",
      "9     490     F      S     F     0\n",
      "10    501     N      Y     N     0\n",
      "11    614     D      G     G     1\n",
      "12    655     H      Y     H     0\n",
      "13    677     Q      H     Q     0\n",
      "14    681     P      R     R     1\n",
      "15    681     P      H     R     0\n",
      "16    701     A      V     A     0\n",
      "FN8970\n",
      "Delta(B.1.617.2)\n",
      "\n",
      "\n",
      "343.0\n",
      "749.0\n",
      "[0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0]\n",
      "   aa_pos aa_WT aa_Mut aa_Fd aa_Ch\n",
      "0     346     R      K     R     0\n",
      "1     417     K      T     K     0\n",
      "2     417     K      N     K     0\n",
      "3     452     L      Q     R     0\n",
      "4     452     L      R     R     1\n",
      "5     477     S      N     S     0\n",
      "6     478     T      K     K     1\n",
      "7     484     E      K     E     0\n",
      "8     484     E      Q     E     0\n",
      "9     490     F      S     F     0\n",
      "10    501     N      Y     N     0\n",
      "11    614     D      G     G     1\n",
      "12    655     H      Y     H     0\n",
      "13    677     Q      H     Q     0\n",
      "14    681     P      R     R     1\n",
      "15    681     P      H     R     0\n",
      "16    701     A      V     A     0\n",
      "FN8979\n",
      "Delta(B.1.617.2)\n",
      "\n",
      "\n",
      "343.0\n",
      "751.0\n",
      "[0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0]\n",
      "   aa_pos aa_WT aa_Mut aa_Fd aa_Ch\n",
      "0     346     R      K     R     0\n",
      "1     417     K      T     K     0\n",
      "2     417     K      N     K     0\n",
      "3     452     L      Q     R     0\n",
      "4     452     L      R     R     1\n",
      "5     477     S      N     S     0\n",
      "6     478     T      K     K     1\n",
      "7     484     E      K     E     0\n",
      "8     484     E      Q     E     0\n",
      "9     490     F      S     F     0\n",
      "10    501     N      Y     N     0\n",
      "11    614     D      G     G     1\n",
      "12    655     H      Y     H     0\n",
      "13    677     Q      H     Q     0\n",
      "14    681     P      R     R     1\n",
      "15    681     P      H     R     0\n",
      "16    701     A      V     A     0\n",
      "FN9008\n",
      "Delta(B.1.617.2)\n",
      "\n",
      "\n",
      "342.0\n",
      "750.0\n",
      "[0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0]\n",
      "   aa_pos aa_WT aa_Mut aa_Fd aa_Ch\n",
      "0     346     R      K     R     0\n",
      "1     417     K      T     K     0\n",
      "2     417     K      N     K     0\n",
      "3     452     L      Q     R     0\n",
      "4     452     L      R     R     1\n",
      "5     477     S      N     S     0\n",
      "6     478     T      K     K     1\n",
      "7     484     E      K     E     0\n",
      "8     484     E      Q     E     0\n",
      "9     490     F      S     F     0\n",
      "10    501     N      Y     N     0\n",
      "11    614     D      G     G     1\n",
      "12    655     H      Y     H     0\n",
      "13    677     Q      H     Q     0\n",
      "14    681     P      R     R     1\n",
      "15    681     P      H     R     0\n",
      "16    701     A      V     A     0\n",
      "FN9014\n",
      "Delta(B.1.617.2)\n",
      "\n",
      "\n",
      "344.0\n",
      "750.0\n",
      "[0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0]\n",
      "   aa_pos aa_WT aa_Mut aa_Fd aa_Ch\n",
      "0     346     R      K     R     0\n",
      "1     417     K      T     K     0\n",
      "2     417     K      N     K     0\n",
      "3     452     L      Q     R     0\n",
      "4     452     L      R     R     1\n",
      "5     477     S      N     S     0\n",
      "6     478     T      K     K     1\n",
      "7     484     E      K     E     0\n",
      "8     484     E      Q     E     0\n",
      "9     490     F      S     F     0\n",
      "10    501     N      Y     N     0\n",
      "11    614     D      G     G     1\n",
      "12    655     H      Y     H     0\n",
      "13    677     Q      H     Q     0\n",
      "14    681     P      R     R     1\n",
      "15    681     P      H     R     0\n",
      "16    701     A      V     A     0\n",
      "FN9035\n",
      "Delta(B.1.617.2)\n",
      "\n",
      "\n",
      "343.0\n",
      "751.0\n",
      "[0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0]\n",
      "   aa_pos aa_WT aa_Mut aa_Fd aa_Ch\n",
      "0     346     R      K     R     0\n",
      "1     417     K      T     K     0\n",
      "2     417     K      N     K     0\n",
      "3     452     L      Q     R     0\n",
      "4     452     L      R     R     1\n",
      "5     477     S      N     S     0\n",
      "6     478     T      K     K     1\n",
      "7     484     E      K     E     0\n",
      "8     484     E      Q     E     0\n",
      "9     490     F      S     F     0\n",
      "10    501     N      Y     N     0\n",
      "11    614     D      G     G     1\n",
      "12    655     H      Y     H     0\n",
      "13    677     Q      H     Q     0\n",
      "14    681     P      R     R     1\n",
      "15    681     P      H     R     0\n",
      "16    701     A      V     A     0\n",
      "FN9037\n",
      "Delta(B.1.617.2)\n",
      "\n",
      "\n",
      "342.0\n",
      "752.0\n",
      "[0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0]\n",
      "   aa_pos aa_WT aa_Mut aa_Fd aa_Ch\n",
      "0     346     R      K     R     0\n",
      "1     417     K      T     K     0\n",
      "2     417     K      N     K     0\n",
      "3     452     L      Q     R     0\n",
      "4     452     L      R     R     1\n",
      "5     477     S      N     S     0\n",
      "6     478     T      K     K     1\n",
      "7     484     E      K     E     0\n",
      "8     484     E      Q     E     0\n",
      "9     490     F      S     F     0\n",
      "10    501     N      Y     N     0\n",
      "11    614     D      G     G     1\n",
      "12    655     H      Y     H     0\n",
      "13    677     Q      H     Q     0\n",
      "14    681     P      R     R     1\n",
      "15    681     P      H     R     0\n",
      "16    701     A      V     A     0\n",
      "FN9068\n",
      "Delta(B.1.617.2)\n",
      "\n",
      "\n",
      "343.0\n",
      "750.0\n",
      "[0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0]\n",
      "   aa_pos aa_WT aa_Mut aa_Fd aa_Ch\n",
      "0     346     R      K     R     0\n",
      "1     417     K      T     K     0\n",
      "2     417     K      N     K     0\n",
      "3     452     L      Q     R     0\n",
      "4     452     L      R     R     1\n",
      "5     477     S      N     S     0\n",
      "6     478     T      K     K     1\n",
      "7     484     E      K     E     0\n",
      "8     484     E      Q     E     0\n",
      "9     490     F      S     F     0\n",
      "10    501     N      Y     N     0\n",
      "11    614     D      G     G     1\n",
      "12    655     H      Y     H     0\n",
      "13    677     Q      H     Q     0\n",
      "14    681     P      R     R     1\n",
      "15    681     P      H     R     0\n",
      "16    701     A      V     A     0\n",
      "FN9077\n",
      "Delta(B.1.617.2)\n",
      "\n",
      "\n",
      "344.0\n",
      "752.0\n",
      "[0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0]\n",
      "   aa_pos aa_WT aa_Mut aa_Fd aa_Ch\n",
      "0     346     R      K     R     0\n",
      "1     417     K      T     K     0\n",
      "2     417     K      N     K     0\n",
      "3     452     L      Q     R     0\n",
      "4     452     L      R     R     1\n",
      "5     477     S      N     S     0\n",
      "6     478     T      K     K     1\n",
      "7     484     E      K     E     0\n",
      "8     484     E      Q     E     0\n",
      "9     490     F      S     F     0\n",
      "10    501     N      Y     N     0\n",
      "11    614     D      G     G     1\n",
      "12    655     H      Y     H     0\n",
      "13    677     Q      H     Q     0\n",
      "14    681     P      R     R     1\n",
      "15    681     P      H     R     0\n",
      "16    701     A      V     A     0\n",
      "FN9081\n",
      "Delta(B.1.617.2)\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Genotipador a partir dos contigs\n",
    "from io import StringIO\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from Bio.Seq import Seq\n",
    "import os\n",
    "import sys\n",
    "\n",
    "#Função para alterar o cabeçalho dos arquivos fasta - coloca no nome do arquivo no lugar do que está escrito no cabeçalho -- isso é necessário pois todos os arquivos tem contig_1 como cabeçalho.\n",
    "def change_fasta_header(outdir_v, indir_v):\n",
    "    import os\n",
    "    import re\n",
    "    \n",
    "    #create output directory\n",
    "    output_dir = outdir_v #\"newheader\"\n",
    "    if not os.path.exists(output_dir):\n",
    "        os.mkdir(output_dir)\n",
    "\n",
    "    #take the path to the data files    \n",
    "    pwd=os.getcwd()+\"/\"+indir_v#\"/data\"\n",
    "    #print(pwd)\n",
    "\n",
    "\n",
    "    #print(pwd)\n",
    "    #for dirname, dirnames, filenames in os.walk(pwd):\n",
    "        #print (dirname)\n",
    "        #print(dirnames)\n",
    "       # print(filenames)\n",
    "\n",
    "    #Get file list inside of data directory.\n",
    "    for file in os.listdir(pwd):    \n",
    "        #if \".empty\" in dirnames:\n",
    "        #    dirnames.remove(\".empty\")\n",
    "        #print (file)    \n",
    "        #print(filefull)\n",
    "\n",
    "        #Select only .nt* or .fa*\n",
    "        if (r'.nt') in file or (r'.fa') in file:\n",
    "            #print(filefull)\n",
    "\n",
    "\n",
    "            myfastalist=list(SeqIO.parse(pwd+\"/\"+file,'fasta'))\n",
    "\n",
    "\n",
    "            for record in myfastalist:\n",
    "                record.id = re.sub(\"[.].*\",\"\",file)\n",
    "                #print(record.id)\n",
    "\n",
    "                with open(outdir_v+\"/\"+file, \"w\") as f_v:\n",
    "                    f_v.write('>'+record.id+'\\n'+str(record.seq)+'\\n')\n",
    "\n",
    "\n",
    "\n",
    "##########################################################\n",
    "#Função para obter uma lista de arquivos e posteriormente alinhar\n",
    "def align_from_files(dir_input, seq_spike):\n",
    "    #make a list of files to be align\n",
    "    ls_f = []\n",
    "    import glob\n",
    "    ls_f = glob.glob(dir_input+\"/*.*\")\n",
    "    #print(ls_f)\n",
    "\n",
    "    #insert Spike sequence\n",
    "    ls_f.insert(0, seq_spike)\n",
    "    #print(ls_f)\n",
    "\n",
    "    align_v = align_f(ls_f)\n",
    "    #print(align_v)\n",
    "    return(align_v)\n",
    "\n",
    "#Função para obter alinhamento a partir de uma lista de arquivos.\n",
    "def align_f(list_v):\n",
    "    from Bio.Align.Applications import MafftCommandline\n",
    "    \n",
    "    file_out1= './output/temp_seqs.fasta'\n",
    "    filenames = list_v\n",
    "    with open(file_out1, 'w') as outfile:\n",
    "        for fname in filenames:\n",
    "            with open(fname) as infile:\n",
    "                for line in infile:\n",
    "                    outfile.write(line)\n",
    "                \n",
    "    mafft_cline = MafftCommandline(\"mafft\", input = file_out1)\n",
    "    stdout,stderr = mafft_cline()\n",
    "\n",
    "    #print(\"Aligment: \\n\" + stdout)\n",
    "    #print (\"\\n\\n Error: \\n\" + stderr)\n",
    "\n",
    "    with open(\"temp_seq_mafft.aln\",\"w\") as hamble:\n",
    "        hamble.write(stdout)\n",
    "        \n",
    "    return(str(stdout))\n",
    "##############################################################\n",
    "\n",
    "\n",
    "#Função para obtenção dos id e sequencias descriminados a partir da variável obtida do alinhamento.\n",
    "def fasta_id_seq_from_align(align_v):\n",
    "    from io import StringIO\n",
    "    fasta_io = StringIO(align_v)\n",
    "    records = SeqIO.parse(fasta_io, \"fasta\")\n",
    "\n",
    "    id_v = []\n",
    "    seq_v = []\n",
    "    for record in records:\n",
    "        #print(\"%s\\t%s\" % (record.id, record.seq[0:10]))\n",
    "        id_v.append(record.id)\n",
    "        seq_v.append(record.seq)\n",
    "    return(id_v, seq_v)\n",
    "    #print(id_v)\n",
    "    #print(seq_v)\n",
    "\n",
    "#Função para carregar as informações presentes na tabela de mutações\n",
    "def load_mutation_tab(file_tb_name_v):\n",
    "    import pandas as pd\n",
    "    df_fe = pd.read_excel(file_tb_name_v, index_col=0)\n",
    "    #print(df_fe)\n",
    "    \n",
    "    #df to dict\n",
    "    dict_mut = {}\n",
    "    for dc_v in range(0,len(df_fe.iloc[:,0])):\n",
    "        if df_fe.index[dc_v] in dict_mut:\n",
    "            dict_mut[df_fe.index[dc_v]].append(list(df_fe.iloc[dc_v,]))\n",
    "        else:\n",
    "            dict_mut[df_fe.index[dc_v]] = list(df_fe.iloc[dc_v,])\n",
    "    #df_fe.iloc[dc_v,]\n",
    "    #df_fe.index[dc_v]\n",
    "    #print(dict_mut)\n",
    "    return(dict_mut)\n",
    " \n",
    "\n",
    "#função para inicio da sequencia após alinhamentos\n",
    "def pos_start_f(seq):\n",
    "    ct = 0\n",
    "    \n",
    "    for char_v in seq:\n",
    "        if(char_v != \"-\"):\n",
    "            #print(ct)\n",
    "            ct_i = ct\n",
    "            break\n",
    "            \n",
    "        ct += 1\n",
    "    \n",
    "    #Caso não tenha nenhum alinhamento e apenas \"-\"\n",
    "    try:\n",
    "        ct_i\n",
    "    except: \n",
    "        ct_i = len(seq)\n",
    "        \n",
    "        \n",
    "    ct = 0\n",
    "    ct_f = 0\n",
    "    #print(seq[ct_i:])\n",
    "    for char_v2 in seq[ct_i:]:\n",
    "        if(char_v2 == \"-\"):\n",
    "            #print(ct)\n",
    "            ct_f = ct\n",
    "            break\n",
    "        ct += 1\n",
    "    \n",
    "    #print(str(ct_i)+\"\\t\"+str(ct_f))\n",
    "    return(ct_i,ct_f)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "#função para obter a posição em frame a traduzir a sequencia corretamente\n",
    "def seq_translate_f(seq_v_i):\n",
    "    #Obter a posição da frame\n",
    "    \n",
    "    start_pos,end_range_pos = pos_start_f(seq_v_i)\n",
    "\n",
    "    if(start_pos > 3):\n",
    "        start_frame_pos = start_pos + (3-start_pos%3)\n",
    "        start_aa_pos = (start_frame_pos/3)+1\n",
    "        if(end_range_pos != 0):\n",
    "            end_pos = start_pos+end_range_pos\n",
    "            end_frame_pos = end_pos-(end_pos%3)\n",
    "            end_aa_pos = (end_frame_pos/3)+1\n",
    "            #print(len(seq_v_i))\n",
    "            #print(start_frame_pos)\n",
    "            #print(end_frame_pos)\n",
    "            #print(end_aa_pos)\n",
    "\n",
    "        else:\n",
    "            end_frame_pos = (len(seq_v_i)-(len(seq_v_i)%3))\n",
    "            end_aa_pos = (end_frame_pos/3)\n",
    "            #print(end_aa_pos)\n",
    "            #print(end_frame_pos)\n",
    "        #print(start_aa_pos)\n",
    "        #print(start_frame_pos)\n",
    "    elif(start_pos == 0):\n",
    "        start_frame_pos = 0\n",
    "        start_aa_pos = 1\n",
    "        if(end_range_pos != 0):\n",
    "            end_pos = start_pos+end_range_pos\n",
    "            end_frame_pos = end_pos-(end_pos%3)\n",
    "            end_aa_pos = (end_frame_pos/3)+1\n",
    "            #print(end_aa_pos)\n",
    "            #print(end_frame_pos)\n",
    "        else:\n",
    "            end_frame_pos = (len(seq_v_i)-(len(seq_v_i)%3))\n",
    "            end_aa_pos = (end_frame_pos/3)\n",
    "            #print(len(seq_v_i))\n",
    "            #print(start_frame_pos)\n",
    "            #print(end_frame_pos)\n",
    "            #print(end_aa_pos)\n",
    "\n",
    "    else:\n",
    "        start_frame_pos = 3\n",
    "        start_aa_pos = 2\n",
    "        if(end_range_pos != 0):\n",
    "            end_pos = start_pos+end_range_pos\n",
    "            end_frame_pos = end_pos-(end_pos%3)\n",
    "            end_aa_pos = (end_frame_pos/3)+1\n",
    "            #print(end_aa_pos)\n",
    "            #print(end_frame_pos)\n",
    "        else:\n",
    "            end_frame_pos = (len(seq_v_i)-(len(seq_v_i)%3))\n",
    "            end_aa_pos = (end_frame_pos/3)\n",
    "\n",
    "\n",
    "    #Obter a sequencia a partir dessa posição.\n",
    "    #seq2 = (seq_v_i)[start_frame_pos:]\n",
    "    #print(seq2)\n",
    "    #if(end_pos != 0):\n",
    "    seq2 = (seq_v_i)[start_frame_pos:end_frame_pos]\n",
    "    #else:\n",
    "    #    seq2 = (seq_v_i)[start_frame_pos:]\n",
    "\n",
    "    #print(seq2)\n",
    "    #print(seq2)\n",
    "    #print(type(seq2))\n",
    "    seq2_tr = seq2.translate()\n",
    "    #print(seq2_tr)\n",
    "    #print(seq2_tr)\n",
    "    return(start_aa_pos, end_aa_pos, seq2_tr)\n",
    "\n",
    "#Função para obter a lista com as mutações encontradas e o dataframe com a descrição\n",
    "def list_df_make_f(mut_list_PTN, seq2_tr,start_aa_pos, end_aa_pos):\n",
    "    #Obter a lista e o DataFrame das mutações encontradas\n",
    "    list_result = []\n",
    "    #print(\"aa_pos\"+\"\\t\"+\"aa_Fd\"+\"\\t\"+\"aa_WT\"+\"\\t\"+\"aa_Mut\")\n",
    "    list_coluns_v = [\"aa_pos\",\"aa_WT\",\"aa_Mut\",\"aa_Fd\",\"aa_Ch\"]\n",
    "    df_result_v = pd.DataFrame()\n",
    "\n",
    "    for coord_v in mut_list_PTN:\n",
    "\n",
    "        #If start of aa sequence was before the first mutation contig\n",
    "        if(int(start_aa_pos) < coord_v[0] and end_aa_pos > coord_v[0]):\n",
    "            aa_F = seq2_tr[int(coord_v[0])-int(start_aa_pos)]\n",
    "            aa_pos = coord_v[0]\n",
    "            aa_wt = coord_v[1]\n",
    "            aa_Mut = coord_v[2]\n",
    "\n",
    "\n",
    "            #print(str(coord_v[0])+\"\\t\"+aa_F+\"\\t\"+coord_v[1]+\"\\t\"+aa_Mut)\n",
    "            #print(list_temp_v)\n",
    "            #print(df_temp_v)\n",
    "\n",
    "            #if a mutation identified... \n",
    "            if(aa_F == aa_Mut):\n",
    "                #List used to check\n",
    "                list_result.append(1)\n",
    "                #DataFrame used in results\n",
    "                list_temp_v = [str(aa_pos),aa_wt,aa_Mut,aa_F,\"1\"]\n",
    "                df_temp_v = pd.DataFrame([list_temp_v], columns=list_coluns_v)\n",
    "                df_result_v = df_result_v.append(df_temp_v,ignore_index=True)\n",
    "\n",
    "\n",
    "            #If no mutation identified\n",
    "            else:\n",
    "                list_result.append(0)\n",
    "\n",
    "                list_temp_v = [str(aa_pos),aa_wt,aa_Mut,aa_F,\"0\"]\n",
    "                df_temp_v = pd.DataFrame([list_temp_v], columns=list_coluns_v)\n",
    "                df_result_v = df_result_v.append(df_temp_v,ignore_index=True)\n",
    "\n",
    "\n",
    "        else:  #If contig was shorter than fisrt mutation position\n",
    "\n",
    "            #aa_F = seq2_tr[int(coord_v[0])-int(start_aa_pos)]\n",
    "            #aa_Mut = coord_v[2]\n",
    "\n",
    "            list_result.append(2)\n",
    "\n",
    "            list_temp_v = [str(coord_v[0]),coord_v[1],coord_v[2],\"NA\",\"NA\"]\n",
    "            df_temp_v = pd.DataFrame([list_temp_v], columns=list_coluns_v)\n",
    "            df_result_v = df_result_v.append(df_temp_v,ignore_index=True)\n",
    "\n",
    "    #print(list_result)\n",
    "    #print(df_result_v)\n",
    "    return(list_result, df_result_v)\n",
    "\n",
    "#Test to find if there is a match\n",
    "#Função para correlacionar as listas\n",
    "def teste_list_f(list_test):\n",
    "    for key_v in dict_mut_v:\n",
    "        if list_test == dict_mut_v[key_v]:\n",
    "            #print(key_v)\n",
    "            #print(df_result_v)\n",
    "            return(key_v)\n",
    "            break\n",
    "\n",
    "            \n",
    "def create_dir_or_del_files (folder_v):\n",
    "    import os, shutil\n",
    "    \n",
    "    if not os.path.exists(folder_v):\n",
    "        os.makedirs(folder_v)\n",
    "    else:\n",
    "        folder = folder_v\n",
    "        for filename in os.listdir(folder):\n",
    "            file_path = os.path.join(folder, filename)\n",
    "            try:\n",
    "                if os.path.isfile(file_path) or os.path.islink(file_path):\n",
    "                    os.unlink(file_path)\n",
    "                elif os.path.isdir(file_path):\n",
    "                    shutil.rmtree(file_path)\n",
    "            except Exception as e:\n",
    "                print('Failed to delete %s. Reason: %s' % (file_path, e))\n",
    "\n",
    "#main  \n",
    "\n",
    "dir_initial_data = \"data6_Placa20211007\"\n",
    "dir_data = \"fasta_temp\"\n",
    "\n",
    "dir_out_v = \"Placa_20211020\"\n",
    "out_file_v = 'Genotipagem_20211021.xlsx'\n",
    "out_dir_file_v = dir_out_v+\"/\"+out_file_v \n",
    "\n",
    "#Cria diretório para fasta ou deleta arquivos no diretório de saída se houver.\n",
    "create_dir_or_del_files(dir_data)\n",
    "\n",
    "#Cria diretório de saída se não existir:\n",
    "if not os.path.exists(dir_out_v):\n",
    "    os.makedirs(dir_out_v)\n",
    "\n",
    "#Correção do cabeçalho fasta de contig_1 para o nome do arquivo.\n",
    "change_fasta_header(dir_data, dir_initial_data)   \n",
    "\n",
    "\n",
    "#Construção do fasta e alinhamento.\n",
    "#dir_data = \"newheader\"\n",
    "ref_seq_file = \"Spike_SeqNucl_LineageB_Wuhan_EPI_ISL_402123_2019.fasta\"\n",
    "align_v = align_from_files(dir_data, ref_seq_file)\n",
    "#print(align_v)\n",
    "\n",
    "#Obtendo id e sequencias dos contigs após alinhamento.    \n",
    "id_v,seq_v = fasta_id_seq_from_align(align_v)\n",
    "\n",
    "#Carregar as informações presentes na tabela de variantes/mutações\n",
    "#Caso tenha alteração nessa tabela não esquecer de alterar também a lista de mutações abaixo.\n",
    "dict_mut_v = load_mutation_tab(\"Tabela_Variant_Mutations_20210909.xlsx\")\n",
    "#print(dict_mut_v)\n",
    "\n",
    "#Lista com as mutações\n",
    "mut_list_PTN = [[346,\"R\",\"K\"],[417,\"K\",\"T\"],[417,\"K\",\"N\"],[452,\"L\",\"Q\"],[452,\"L\",\"R\"],[477,\"S\",\"N\"],[478,\"T\",\"K\"],[484,\"E\",\"K\"],[484,\"E\",\"Q\"],[490,\"F\",\"S\"],[501,\"N\",\"Y\"],[614,\"D\",\"G\"],[655,\"H\",\"Y\"],[677,\"Q\",\"H\"],[681,\"P\",\"R\"],[681,\"P\",\"H\"],[701,\"A\",\"V\"]]\n",
    "#print([i[0] for i in mut_list_PTN[0:]])\n",
    "#print(len(mut_list_PTN))\n",
    "\n",
    "#Fazer loop aqui para bustar todos os registros\n",
    "#Obtendo sequencia de aa\n",
    "\n",
    "#Carregando arquivo a ser escrito no loop\n",
    "#writer_f = pd.ExcelWriter('testFile.xlsx', engine='xlsxwriter')\n",
    "\n",
    "df_summary_out = pd.DataFrame()\n",
    "dic_out_print = {}\n",
    "\n",
    "ct_seq_v = 1\n",
    "for seq_nc_v in seq_v[1:]:\n",
    "    #print(seq_nc_v)\n",
    "    st_aa_pos, ed_aa_pos, seq2_tr = seq_translate_f(seq_nc_v)\n",
    "    print(st_aa_pos)\n",
    "    print(ed_aa_pos)\n",
    "    #print(seq2_tr)\n",
    "\n",
    "    #Obtendo list de mutações e descrição\n",
    "    ls_temp,df_temp_res = list_df_make_f(mut_list_PTN, seq2_tr, st_aa_pos, ed_aa_pos)\n",
    "    print(ls_temp)\n",
    "    print(df_temp_res)\n",
    "\n",
    "    #Obtendo correlação com lista de variantes\n",
    "    result_variant_v = teste_list_f(ls_temp)\n",
    "    print(id_v[ct_seq_v])\n",
    "    print(str(result_variant_v)+\"\\n\\n\")\n",
    "    \n",
    "    df_summary_out = pd.concat([df_summary_out,pd.DataFrame({\"Sample\":[id_v[ct_seq_v]],\"Variant Name\":[result_variant_v]})])\n",
    "    #Escrevendo para o arquivo detalhado.\n",
    "    df_out0 = pd.DataFrame({\"aa_pos\":[\"Sample Name:\"],\"aa_WT\":[id_v[ct_seq_v]]})\n",
    "    df_out1 = pd.DataFrame({\"aa_pos\":[\"Variant Name:\"],\"aa_WT\":[result_variant_v]})\n",
    "    df_out_print = pd.concat([df_temp_res,df_out0,df_out1])\n",
    "    #dic_out_print[id_v[ct_seq_v]] = [df_out_print,result_variant_v]\n",
    "    dic_out_print[id_v[ct_seq_v]] = df_out_print\n",
    "    \n",
    "    #df_out_print.to_excel(writer_f,sheet_name=(id_v[ct_seq_v])[0:10], index = False)\n",
    "    #result_variant_v.to_excel(writer_f,sheet_name=(id_v[ct_seq_v])[0:10])\n",
    "    \n",
    "    ct_seq_v+=1\n",
    "\n",
    "\n",
    "#writer_f.save()\n",
    "#writer_f.close()\n",
    "\n",
    "#print(df_summary_out)    \n",
    "#print(dic_out_print)\n",
    "\n",
    "#Imprimindo arquivo de saida:\n",
    "\n",
    "writer_out_f = pd.ExcelWriter(out_dir_file_v, engine='xlsxwriter')\n",
    "df_summary_out.to_excel(writer_out_f,sheet_name=\"SUMMARY\", index = False)\n",
    "for key_v in dic_out_print:\n",
    "    #(dic_out_print[key_v])[0].to_excel(writer_out_f,sheet_name=key_v[0:20], startrow = 0, startcol = 0, index = False)\n",
    "    dic_out_print[key_v].to_excel(writer_out_f,sheet_name=key_v[0:20], index = False)\n",
    "    \n",
    "    #worksheet = writer.sheets[]\n",
    "    #worksheet.write(0, 0, \"Id_Sample=\"+str(key_v))\n",
    "    #worksheet.write(1, 0, \"Variant Name=\"+str((dic_out_print[key_v])[1]))\n",
    "writer_out_f.save()\n",
    "                        \n",
    "                        \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "a931b407",
   "metadata": {},
   "outputs": [],
   "source": [
    "def del_files_from_folder_f (folder_v):\n",
    "    import os, shutil\n",
    "    folder = folder_v\n",
    "    for filename in os.listdir(folder):\n",
    "        file_path = os.path.join(folder, filename)\n",
    "        try:\n",
    "            if os.path.isfile(file_path) or os.path.islink(file_path):\n",
    "                os.unlink(file_path)\n",
    "            elif os.path.isdir(file_path):\n",
    "                shutil.rmtree(file_path)\n",
    "        except Exception as e:\n",
    "            print('Failed to delete %s. Reason: %s' % (file_path, e))\n",
    "\n",
    "del_files_from_folder_f(\"newheader\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05381068",
   "metadata": {},
   "source": [
    "# Alterações para a proxima versão:\n",
    "- ERRO IDENTIFICADO!!!, caso haja alguma deleção as sequencias seguintes ficam com problema.\n",
    "- Deletar os diretórios de cabeçalho automaticamente, armazenar os arquivos gerados automaticamente.\n",
    "- GUI ou alternativa para manipulação mais fácil\n",
    "- Screening de todas as mutações no contig\n",
    "- Acréscimo de mais mutações de interesse a serem verificadas também de maneira pontual.\n",
    "- Apontamento das novas mutações para checagem no espectro... (checagem remota do espectro?)\n",
    "- Probabilidade de mutações em cada posição para a genotipagem mais precisa??? (entender melhor o que o PANGOLIM faz)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
